File: algorithm_python/Algorithm_Session/Week4/boj_1012.py
import sys

input = sys.stdin.readline
sys.setrecursionlimit(10000)

def dfs(graph, i, j):
    if i < 0 or j < 0 or i > len(graph) or j > len(graph[i]):
        return False

    if graph[i][j] == 1:
        graph[i][j] = 0

        dfs(graph, i + 1, j)
        dfs(graph, i - 1, j)
        dfs(graph, i, j + 1)
        dfs(graph, i, j - 1)

        return True
    
    return False


t = int(input())

while t:
    cols, rows, k = map(int, input().split())
    graph = [[0] * (cols + 1) for _ in range(rows + 1)]

    for _ in range(k):
        x, y = map(int, input().split())
        graph[y][x] = 1

    cnt = 0
    for i in range(rows):
        for j in range(cols):
            if dfs(graph, i, j):
                cnt += 1
    
    print(cnt)

    t -= 1

File: 02_Algorithm/algorithm_java/Week4/boj_1012.java
package Week4;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class boj_1012 {
    private static StringTokenizer st;
    private static int[][] graph;
    private final static BufferedReader br =
            new BufferedReader(new InputStreamReader(System.in));

    private static boolean dfs(int x, int y) {
        if (x < 0 || y < 0 || x >= graph.length || y >= graph[x].length)
            return false;

        if (graph[x][y] == 1) {
            graph[x][y] = 0;

            dfs(x + 1, y);
            dfs(x - 1, y);
            dfs(x, y - 1);
            dfs(x, y + 1);

            return true;
        }

        return false;
    }

    public static void main(String[] args) throws IOException {
        int t = Integer.parseInt(br.readLine());

        while (t > 0) {
            st = new StringTokenizer(br.readLine());

            int cols = Integer.parseInt(st.nextToken());
            int rows = Integer.parseInt(st.nextToken());
            int k = Integer.parseInt(st.nextToken());
            graph = new int[rows + 1][cols + 1];

            for (int i = 0; i < k; i++) {
                st = new StringTokenizer(br.readLine());
                int x = Integer.parseInt(st.nextToken());
                int y = Integer.parseInt(st.nextToken());

                graph[y][x] = 1;
            }

            int cnt = 0;

            for (int i = 0; i < rows; i++)
                for (int j = 0; j < cols; j++)
                    if (dfs(i, j))
                        cnt++;

            System.out.println(cnt);
            t--;
        }
    }
}
